File: image_utils.py
import cv2 as cv
import numpy as np
import random

def rotate_img(img, degree):
    cols, rows = img.shape 
    M = cv.getRotationMatrix2D((cols/2, rows/2), degree, 1)
    dst = cv.warpAffine(img, M, (cols, rows))
    return dst

def translate_img(img, x_trans, y_trans):
    cols, rows = img.shape 
    M = np.float32([[1, 0, x_trans],[0, 1, y_trans]])
    dst = cv.warpAffine(img, M, (cols,rows))
    return dst

def add_random_noise(img, strength):
    cols, rows = img.shape 
    num = random.randint(0, strength)
    for i in range(num):
        col = random.randint(0, cols-1)
        row = random.randint(0, rows-1)
        noise = random.randint(-30, 30)
        value = img[col][row] + noise
        value = min(value, 255)
        value = max(0, value)
        img[col][row] = value
    return img

File: ImageUtils.cpp
#include "ImageUtils.hpp"

using namespace std;

void rotateImage(cv::Mat& img, float degree) {
    cv::Point2f center(cols/2, rows/2);
    cv::Mat rotateMatrix = cv::getRotationMatrix2D(center, degree, 1);

    cv::warpAffine(img, img, rotateMatrix, image.size());
}

void translateImage(cv::Mat& img, int offsetX, int offsetY) {
    cv::Mat transMatrix = (cv::Mat_<float>(2,3) << 1, 0, offsetX, 0, 1, offsetY);

    cv::warpAffine(img, img, transMatrix, img.size());
}

void addRandomNoise(cv::Mat& img, int strength) {
    int cols = img.cols;
    int rows = img.rows;
    cv::Mat_ cimg = img;

    for (int i = 0; i < strength; i++) {
        int col = rand() % cols;
        int row = rand() % rows;
        int noise = rand() % 61 - 30;
        float value = cimg(col, row) + noise;
        value = min(value, 255);
        value = max(0, value);
        cimg(col, row) = value;
    }
}